framework/Optimizers/Conjugate.py

```diff
@@ -0,0 +1,421 @@
+              self.raiseAMessage(' ... Trajectory "{}" converged {} times consecutively!'.format(traj,self.counter['persistence'][traj]))
+              self.convergeTraj[traj] = True
+              self.removeConvergedTrajectory(traj)
+            else:
+              self.raiseAMessage(' ... converged Traj "{}" {} times, required persistence is {}.'.format(traj,self.counter['persistence'][traj],self.convergencePersistence))
+
+          self.counter['alpha'][traj] = self.counter['lastStepSize'][traj]
+
+          grad = dict((var,self.counter['gfk'][traj][ind]/(self.optVarsInit['upperBound'][var]-self.optVarsInit['lowerBound'][var])) for ind,var in enumerate(self.getOptVars()))
+
+          try:
+            self.counter['gradNormHistory'][traj][1] = self.counter['gradNormHistory'][traj][0]
+          except IndexError:
+            pass # don't have a history on the first pass
+          self.counter['gradNormHistory'][traj][0] = grad
+
+          new = self._newOptPointAdd(grad, traj)
+          if new is not None:
+            # add new gradient points
+            self._createPerturbationPoints(traj, new)
+          # reset storage
+          self._setupNewStorage(traj)
+
+
+  ###################
+  # Utility Methods #
+  ###################
+
+  def _getPerturbationDirection(self, perturbationIndex, step = None):
+    """
+      This method is aimed to get the perturbation direction (i.e. in this case the random perturbation versor)
+      @ In, perturbationIndex, int, the perturbation index (stored in self.perturbationIndices)
+      @ In, step, int, the step index, zero indexed
+      @ Out, direction, list, the versor for each optimization dimension
+    """
+    _, varId, denoId, cdId = self._identifierToLabel(perturbationIndex)
+    direction = np.zeros(len(self.getOptVars())).tolist()
+    if cdId == 0:
+      direction[varId] = 1.0
+    else:
+      direction[varId] = -1.0
+    if step:
+      if step % 2 == 0:
+        factor = 1.0
+      else:
+        # flip the sign of the direction, this step will not affect central differancing
+        # but will make the direction between forward and backward
+        # for example of 2 variables 3 denoise W/O central diff:
+        # step 0 directions are ([1 0],[0 1])*3
+        # step 1 directions are ([-1 0],[0 -1])*3
+        # with central diff:
+        # step 0 directions are ([1 0],[0 1],[-1,0],[0,-1])*3
+        # step 1 directions are ([-1,0],[0,-1],[1 0],[0 1])*3
+        factor = -1.0
+      direction = [var * factor for var in direction]
+
+    self.currentDirection = direction
+    return direction
+
+  def localEvaluateGradient(self, traj):
+    """
+      Local method to evaluate gradient.
+      @ In, traj, int, the trajectory id
+      @ Out, gradient, dict, dictionary containing gradient estimation. gradient should have the form {varName: gradEstimation}
+    """
+    gradient = {}
+    gi = {}
+    inVars = self.getOptVars()
```

Comment:
How much of this method is common to the other Optimizers? For what I see (but I might be mistaken), quite a lot

Reply:
_getPerturbationDirection is the same with Finite difference, but different from base class(SPSA).
 localEvaluateGradient is not the same for every method, in the conjugate gradient, it updates the gradient history and tries to store the best value. Other methods seem to update gradient history in checking convergence step.

Reply:
Change the base class from SPSA into finite-difference.